### DISCUSSION TITLE
How to rewrite lsp keymaps for certain lang only?

### DISCUSSION THREAD

**Discussion by kapral18** [LazyVim/LazyVim#3960] [Q&A] [answered]
I want to add `<leader>cR` for ts,js files only, but it's an lsp keymap. I can remove it in lspconfig opts like said in README but to overwrite with tstools keymaps, but that will leave them permanently disabled for session, how can I make sure the default mapping from lazyvim gets applied in all other situaions?

```lua
  {
    "pmizio/typescript-tools.nvim",
    keys = {
      { "gD", ft = ft_js, "<cmd>TSToolsGoToSourceDefinition<cr>", desc = "TSTools: Goto Source Definition" },
    },
    config = true
  },
```

**Comment by dpetka2001** [comment]
It should just overwrite it, no? I just tested with `{ "<leader>cR", "<cmd>echo 'hello'<cr>", ft = "json" }` and in Lua files it says `Rename File` and in json files it just prints "hello".

**Reply by kapral18** [reply]
hmm I thought lsp keymaps can only be overwritten via  `nvim-lspconfig` https://www.lazyvim.org/plugins/lsp#%EF%B8%8F-customizing-lsp-keymaps

```lua
          local keys = require("lazyvim.plugins.lsp.keymaps").get()

          keys[#keys + 1] = { "gD", false or new value }
```

**Reply by kapral18** [reply]
anyways in my case lsp keys take over keys={} defined mappings

**Reply by dpetka2001** [reply]
They're just keymaps. I think it's natural that you can overwrite them. It's just that LSP keymaps are only available when a LSP server is attached. The way in the docs is if you want to disable/change the LSP action with some other LSP action for some mapping. I just wasn't sure if the `keys` that also have a `ft` set would get registered after the LSP keymaps or not,but it seems that it works just fine. Just try it out to see if it works for you.

**Reply by dpetka2001** [reply, marked as answer]
Ok I was wrong in my previous comment because I didn't have a json LSP server attached to the buffer that i tested, so the LSP mappings were not in effect and that's why it worked by mistake. You could try the following
```lua
  {
    "neovim/nvim-lspconfig",
    opts = function()
      local Keys = require("lazyvim.plugins.lsp.keymaps").get()

      vim.api.nvim_create_autocmd("LspAttach", {
        group = vim.api.nvim_create_augroup("js_keys", {}),
        desc = "JS keys",
        callback = function(ev)
          if vim.bo[ev.buf].filetype == "javascript" then
            vim.list_extend(Keys, {
              {
                "<leader>cR",
                false,
              },
            })
          else
            vim.list_extend(Keys, {
              {
                "<leader>cR",
                LazyVim.lsp.rename_file,
                desc = "Rename File",
                mode = { "n" },
                has = { "workspace/didRenameFiles", "workspace/willRenameFiles" },
              },
            })
          end
        end,
      })
    end,
  },
```
and see if it works for you as you'd like. I tested and it seems to do what you want, but I might have missed something, so better if you test it as well. There might be another better way, but I can't think of it.

**Reply by kapral18** [reply]
yes this is exactly what I need thank you @dpetka2001.

**Reply by kapral18** [reply]
Based on your idea but a little differently without redefining lazyvim keys explicitly
https://github.com/kapral18/dotfiles/blob/9a2fcdf3791ac8bc28183fb5b748eb9e337b991c/home/dot_config/exact_nvim/exact_lua/exact_plugins/typescript.lua#L81-L138

Thanks again

**Reply by dpetka2001** [reply]
Cool implementation!! Thanks for sharing and letting me expand my knowledge :smile:

**Reply by firubat** [reply]
@kapral18 @dpetka2001 In these solutions, are the keymaps changed globaly or only in the respective filetype buffers? That is, if for example you open both a json and a lua file in the same session, will each have their own keymappings?
